sklearn/decomposition/_dict_learning.py

```diff
@@ -1844,37 +2219,110 @@ def fit(self, X, y=None):
             Returns the instance itself.
         """
-        random_state = check_random_state(self.random_state)
-        X = self._validate_data(X)
+        self._batch_size = self.batch_size
+        if self.batch_size == "warn":
+            warnings.warn(
+                "The default value of batch_size will change from 3 to 256 in 1.3.",
+                FutureWarning,
+            )
+            self._batch_size = 3
 
-        U, (A, B), self.n_iter_ = dict_learning_online(
-            X,
-            self.n_components,
-            alpha=self.alpha,
-            n_iter=self.n_iter,
-            return_code=False,
-            method=self.fit_algorithm,
-            method_max_iter=self.transform_max_iter,
-            n_jobs=self.n_jobs,
-            dict_init=self.dict_init,
-            batch_size=self.batch_size,
-            shuffle=self.shuffle,
-            verbose=self.verbose,
-            random_state=random_state,
-            return_inner_stats=True,
-            return_n_iter=True,
-            positive_dict=self.positive_dict,
-            positive_code=self.positive_code,
+        X = self._validate_data(
+            X, dtype=[np.float64, np.float32], order="C", copy=False
         )
-        self.components_ = U
-        # Keep track of the state of the algorithm to be able to do
-        # some online fitting (partial_fit)
-        self.inner_stats_ = (A, B)
-        self.iter_offset_ = self.n_iter
-        self.random_state_ = random_state
+
+        self._check_params(X)
+        self._random_state = check_random_state(self.random_state)
+
+        dictionary = self._initialize_dict(X, self._random_state)
+        dict_buffer = dictionary.copy()
+
+        if self.shuffle:
+            X_train = X.copy()
+            self._random_state.shuffle(X_train)
+        else:
+            X_train = X
+
+        n_samples, n_features = X_train.shape
+
+        if self.verbose:
+            print("[dict_learning]")
+
+        # Inner stats
+        self._inner_stats = (
+            np.zeros((self._n_components, self._n_components), dtype=X_train.dtype),
+            np.zeros((n_features, self._n_components), dtype=X_train.dtype),
+        )
+
+        if self.max_iter is not None:
+
+            # Attributes to monitor the convergence
+            self._ewa_cost = None
+            self._ewa_cost_min = None
+            self._no_improvement = 0
+
+            batches = gen_batches(n_samples, self._batch_size)
+            batches = itertools.cycle(batches)
+            n_steps_per_epoch = int(np.ceil(n_samples / self._batch_size))
+            n_steps = self.max_iter * n_steps_per_epoch
+
+            # allow max_iter = 0
+            i = -1
+
+            for i, batch in zip(range(n_steps), batches):
+                this_X = X_train[batch]
+
+                batch_cost = self._minibatch_step(
+                    this_X, dictionary, self._random_state, i
+                )
+
+                if self._minibatch_convergence(
```

Comment:
Reading this line, I am thinking that having a name such as `self._has_converged_...` could be more readable.

Reply:
I went for ``_check_convergence`` because it does a little more than just returning a bool